### DISCUSSION TITLE
Renovate fails to make digest updates unless GOPROXY contains `direct`

### DISCUSSION THREAD

**Discussion by colinodell** [renovatebot/renovate#24577] [Request Help] [answered]
### How are you running Renovate?

Self-hosted Renovate

### If you're self-hosting Renovate, select which platform you are using.

GitLab self-hosted

### Was this something which used to work for you, and then stopped?

I am trying to get this working for the first time

### Describe the problem

We have a user who is experiencing `Failed to look up go package golang.org/x/exp` warnings instead of receiving the expected digest bumps.  Notably:

- `golang.org/x/exp` has no tags
- We have `GOPROXY=https://proxy.golang.org` set

Under the hood, `GoDatasource.getReleases()` will call `GoProxyDatasource.getReleases()` and/or `GoDirectDatasource.getReleases()` depending on how `GOPROXY` is set.  For go modules without tags (like this one), both may return an object with an empty `releases` array:

```
// GoProxyDatasource.getReleases() returns:
{
  releases: [],
  tags: {
    latest: "v0.0.0-20230905200255-921286631fa9"
  }
}

// GoDirectDatasource.getReleases() returns:
{
  sourceUrl: null,
  releases: []
}
```

However, the final result from `GoDatasource` differs based on the contents of `GOPROXY`:

| `GOPROXY` | Returned value |
| --- | --- |
| `https://proxy.golang.org` | `null` |
| `direct` | `{ sourceUrl: null, releases: [] }`
| `https://proxy.golang.org,direct` | `{ sourceUrl: null, releases: [] }`

Basically, results from `GoProxyDatasource` are **always discarded** if `releases == []`, but results from `GoDirectDatasource` are always returned immediately - even if `releases` is empty. Presumably this is because we don't know if the next proxy in the list (or `direct`) might contain `releases`.  However, this logic means that `tags.latest` is discarded, even in cases where no subsequent proxies or `direct` will be checked.  This has the net effect of the datasource acting as if the dependency just doesn't exist.

Perhaps this behavior should be changed such that empty results from a proxy may be returned, but only if no subsequent sources find releases?  This would effectively result in this behavioral change:

| `GOPROXY` | Old returned value | New behavior |
| --- | --- | --- |
| `https://proxy.golang.org` | `null` |  **`{ releases: [], tags:{latest: "v0.0.0-20230905200255-921286631fa9"}}`**
| `direct` | `{ sourceUrl: null, releases: [] }` | `{ sourceUrl: null, releases: [] }` (unchanged)
| `https://proxy.golang.org,direct` | `{ sourceUrl: null, releases: [] }` | `{ sourceUrl: null, releases: [] }` (unchanged)

And thus Renovate would return (instead of discard) the knowledge about that latest digest.

### Relevant debug logs

<details><summary>Logs</summary>

```
DEBUG: GET https://proxy.golang.org/golang.org/x/exp/v2/@v/list = (code=ERR_NON_2XX_3XX_RESPONSE, statusCode=404 retryCount=0, duration=273) (repository=foo/bar)
DEBUG: Failed to look up go package golang.org/x/exp (repository=foo/bar, packageFile=go.mod, dependency=golang.org/x/exp)
```

</details>

<details><summary>Relevant http statistics</summary>

```
         "https://proxy.golang.org/golang.org/x/exp/@latest (GET,200)": 1,
         "https://proxy.golang.org/golang.org/x/exp/@v/list (GET,200)": 1,
         "https://proxy.golang.org/golang.org/x/exp/v2/@v/list (GET,404)": 1,
```

</details>

### Have you created a minimal reproduction repository?

I have explained in the description why a minimal reproduction is impossible

**Comment by rarkins** [comment]
I think it sounds ok, but I'm not certain. How will you _use_ that data? Is it just to show that there wasn't a lookup error?

**Reply by colinodell** [reply]
> Is it just to show that there wasn't a lookup error?

That's part of it.  Those lookup errors occur when `null` results (which [get turned into `no-result` lookup errors](https://github.com/renovatebot/renovate/blob/bb6bc356dcae5753197a093a99466f043b07b8bd/lib/modules/datasource/index.ts#L349-L349)) are handled [in this part of the lookup process](https://github.com/renovatebot/renovate/blob/bb6bc356dcae5753197a093a99466f043b07b8bd/lib/workers/repository/process/lookup/index.ts#L93-L108).  But not only does that log a warning, it **also aborts the lookup process before reaching the code to [update the existing digest](https://github.com/renovatebot/renovate/blob/bb6bc356dcae5753197a093a99466f043b07b8bd/lib/workers/repository/process/lookup/index.ts#L423-L429)**!

So looking at the resulting behavior in those three scenarios:

| `GOPROXY` | Returned value | Net Result |
| --- | --- | --- |
| `https://proxy.golang.org` | `null` | Lookup short-circuits before digest is added 😢  |
| `direct` | `{ sourceUrl: null, releases: [] }` | Lookup proceeds down to adding the digest ✅   |
| `https://proxy.golang.org,direct` | `{ sourceUrl: null, releases: [] }` | Lookup proceeds down to adding the digest ✅   |

I basically want to make the first scenario to work like the other two: **if the datasource knows that the dependency exists, it should not return `null`**.

So we need to return something that is [neither `null` nor just `{ releases: [] }`](https://github.com/renovatebot/renovate/blob/bb6bc356dcae5753197a093a99466f043b07b8bd/lib/modules/datasource/index.ts#L306-L308), as both of those values are used to signal `no-result`.  Thankfully the proxy already includes the latest known version in these cases, so we'll pass that data as an indicator that the dependency actually does exist with a known digest despite not having tags.

**Reply by colinodell** [reply, marked as answer]
I've opened PR #24578 with the proposed changes.

**Comment by alock** [comment]
I believe this change can result in issues of timeout for self-hosted renovate.  Our renovate runs in a no-public-egress network but we override all proxies to our private mirrors.  When the repository has `golang.org/x` dependencies that do not have semver it will try to reach out to the internet via `https://golang.org/x/image?go-get=1`.  With the GOPROXY overridden correct it finds non-semver upadates but still tries to gather more data for some reason via the internet path.  I'm not that experienced with the repository structure so I couldn't find the fix, but wanted to post to let people know if they bump into the issue too.

My workaround was to migrate all `golang.org/x` dependencies to latest or at a minimum the semver release.
